src/Microsoft.Graph.Core/Tasks/LargeFileUploadTask.cs

```diff
@@ -0,0 +1,258 @@
+                    catch (ServiceException exception)
+                    {
+                        if (exception.IsMatch("generalException") || exception.IsMatch("timeout"))
+                        {
+                            if (firstAttempt)
+                            {
+                                firstAttempt = false;
+                                exceptionTrackingList.Add(exception);
+                            }
+                            else
+                            {
+                                throw;
+                            }
+                        }
+                        else if (exception.IsMatch("invalidRange"))
+                        {
+                            // Succeeded previously, but nothing to return right now
+                            return new UploadResult<T>();
+                        }
+                        else
+                        {
+                            throw;
+                        }
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Get the series of requests needed to complete the upload session. Call <see cref="UpdateSessionStatusAsync"/>
+        /// first to update the internal session information.
+        /// </summary>
+        /// <param name="options">Options to be applied to each request.</param>
+        /// <returns>All requests currently needed to complete the upload session.</returns>
+        internal IEnumerable<UploadSliceRequest<T>> GetUploadSliceRequests(IEnumerable<Option> options = null)
+        {
+            foreach (var (item1, item2) in this._rangesRemaining)
+            {
+                var currentRangeBegins = item1;
+
+                while (currentRangeBegins <= item2)
+                {
+                    var nextSliceSize = NextSliceSize(currentRangeBegins, item2);
+                    var uploadRequest = new UploadSliceRequest<T>(
+                        this.Session.UploadUrl,
+                        this._client,
+                        options,
+                        currentRangeBegins,
+                        currentRangeBegins + nextSliceSize - 1,
+                        this.TotalUploadLength);
+
+                    yield return uploadRequest;
+
+                    currentRangeBegins += nextSliceSize;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Upload the whole session.
+        /// </summary>
+        /// <param name="maxTries">Number of times to retry entire session before giving up.</param>
+        /// <param name="options">Query and header option name value pairs for the request.</param>
+        /// <param name="progress">IProgress object to monitor the progress of the upload.</param>
+        /// <returns>Item information returned by server.</returns>
+        public async Task<UploadResult<T>> UploadAsync(IProgress<long> progress = null, int maxTries = 3, IEnumerable<Option> options = null)
```

Comment:
How does a client pause upload to make it resume able? I'd expect a cancellationToken in both `UploadAsync` and `ResumeAsync` so the client can stop upload and pause the upload gracefully.

Reply:
The idea is not to have the client be able to pause the upload but to be able to recover from the upload in case something trips up during the upload and therefore be able to resume using the uploadTaskObject in hand.
Something like what is described [here](https://github.com/microsoftgraph/msgraph-sdk-javascript/blob/dev/docs/tasks/LargeFileUploadTask.md#we-can-just-resume-the-broken-upload).